### PR TITLE
Restored cache for the ZincScalaCompiler

### LINKED ISSUE #12591: Slow Scala compilation times in 6.2.2 compared to 5.6.2
### Expected Behavior
I expected build times to stay the same or drop after upgrade to the newer version of Gradle. Small performance drop is also acceptable.

### Current Behavior
For multimodule builds with Scala code build times increased dramatically (2 times the usual time). I've traced the problem to the tasks that compile Scala. I think this might be related to the Zinc upgrade.

### Context
Fast compilation: Gradle 5.6.2
Slow compilation: Gradle 6.2.2, 6.3-rc4

I have two projects build with gradle (small one ~ 10 modules, large one ~ 100 modules). The problem occurs for both of them.

For small project:
```
Gradle 5.6.2:

gw clean compileScala --no-build-cache --profile

Two instances of GradleWorkerMain:
 - Non heap memory at the end: 180MB, 202MB
 - Loaded classes: 11274, 11047

Total time:
BUILD SUCCESSFUL in 2m 28s
61 actionable tasks: 60 executed, 1 up-to-date

Gradle 6.2.2:

gw clean compileScala --no-build-cache --profile

Two instances of GradleWorkerMain:
 - Non heap memory at the end: 714MB, 716MB
 - Loaded classes: 44244, 43294

Total time:
BUILD SUCCESSFUL in 4m 29s
61 actionable tasks: 60 executed, 1 up-to-date
```

I've profiled Gradle workers and have noticed a large increase in loaded classes and non-heap space usage. In 6.2.2 a lot of classes are loaded multiple times, for example:
 - scala.tools.nsc.Global (5 instances in different classloaders)

I've tried to build a single module and the times are the same in 5.6.2 and 6.2.2. Only multimodule builds are noticable slower.

Note: I'm not using the incremental Scala compilation:
```
        scalaCompileOptions.with {
            force = true
        }
```

### Steps to Reproduce 
I think this can be reproduced by any multimodule build with scala code. 
Here is an example:
https://github.com/SheliakLyr/gradle-scala-slow-compilation-example

I tried to build it with 5.6.2 and 6.2.2: `./gradlew clean build --scan`.
The build scan for 5.6.2: https://scans.gradle.com/s/dy5hsgpq3dotw
The build scan for 6.2.2: https://scans.gradle.com/s/ujdltbw6csltc

Task execution time for 5.6.2: 28s
Task execution time for 6.2.2: 6m+

### Your Environment
https://scans.gradle.com/s/ujdltbw6csltc
I will be happy to answer any questions that might help you solve this issue.

### PR DESCRIPTION
**Pull request by SheliakLyr**
Reverts part of 55cc646
Issue: #12591

Signed-off-by: Sheliak Lyr <sheliak.lyr@gmail.com>

Fixes #12591 

### Context
Fixes performance regression introduced in version 6.0. See Issue #12591 for details.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
@big-guy At first sight, I can't see how that cache breaks the reason for removing it in the first place. But I may be missing something.
Any opinion?

**Comment by big-guy** [comment]
@ljacomet I'm +1 for fixing this.  I should have added a perf test when I made this switch before. 

I'm -1 with fixing it in this way.  There are a few problems:
- The cache isn't limited in any way or cleaned up.  Workers are reused across subprojects, so this could impact large multi-project builds.
- The cache is based on the file paths and not the contents of the classpath. The classpath can be made up of changing jars with identical paths (although, most of the time it isn't).
- We already cache most of the worker's implementation, so caching the ZincScalaCompiler is masking something else. 
- I'm not sure we can safely reuse ZincScalaCompiler due to the way we use AnalysisStoreProvider.  There's certainly some state that would get carried across subprojects now.

I believe the actual problem is this classloader is not reused/cached: https://github.com/gradle/gradle/blob/4d87cd9a24fb9070f683d5b603e3606da01b38b7/subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java#L85

By caching the ZincScalaCompiler, we also reuse this classloader.  If you change that code to...

```
...
    static ClassLoader scalaClassLoader;

    private static ScalaInstance getScalaInstance(final Iterable<File> scalaClasspath) {
        if (scalaClassLoader == null) {
            scalaClassLoader = getClassLoader(scalaClasspath);
        }
...
```

I see a similar level of improvement.  This is obviously the wrong kind of fix too, but I think it demonstrates that reusing that classloader has some impact. I suspect this may be something that the old zinc compiler was doing for us since we didn't worry about caching these things before.

With 6.4-rc-2: https://e.grdev.net/s/cceccmm6swwxg/performance/execution 38 minutes of task execution
With naive fix above: https://e.grdev.net/s/wficag6rkd4iu/performance/execution 21 minutes of task execution

(sorry, these are private links)

I think a proper fix would be to limit the change to just caching this classloader.  It would be nice if we had access to some of the infrastructure we have for this (e.g., `HashingClassLoaderFactory` and `ClassLoaderCache`), but those aren't available in the worker and getting those services into the worker is a bigger task. 

Maybe we can do something a little simpler:
- Instead of sending a `Set<File>` to `ZincScalaCompilerFactory#getCompiler`, let's send a `ClassPath` and `HashCode`.
- On the daemon side (where we have all the services we need), calculate the `HashCode` for the Scala `ClassPath` with `ClasspathHasher` in `ScalaCompilerFactory` and `DefaultScalaToolProvider` to pass through to `ZincScalaCompilerFactory`.
- In `ZincScalaCompilerFactory`, use a Guava cache with weak values or a fixed size to cache `HashCode` -> `ClassLoader`.  

WDYT?

**Comment by SheliakLyr** [comment]
I agree with all your points. I think I can prepare a better fix according to this guidelines. Unless someone else wants to tackle the issue.

**Comment by ljacomet** [comment]
@SheliakLyr If you are willing to keep working on this, please do so. Otherwise I can't guarantee it will be done in the short term.
Thanks a lot for your work so far!

**Comment by SheliakLyr** [comment]
I've just pushed a new commit. 

| Version  | Scan | Time |
| ------------- | ------------- | ------------- |
| 5.6.2  |  https://scans.gradle.com/s/5chjimqjtjgjs | 29s |
| 6.0.0 |  https://scans.gradle.com/s/tthgfl4smebhw | 6m19s |
| FixV1 (naive, previous)  |  https://gradle.com/s/5yxhz5i7ndjzq | 26s |
| *FixV2 (cached CL, current)* |  https://scans.gradle.com/s/vpi4kcadul4ei | 43s |
| FixV4 (V2 + cache in analyzing compiler)  |  https://scans.gradle.com/s/ln2y3e3use6a6 | 26s |

Currently, this branch contains "FixV2". Unfortunately performance is still worse than 5.6.x, but compared to 6.x improvement is significant! Keep in mind that my test-project is specifically designed to showcase the problem.

In branch cachedScalaCompilerV4 you can see an additional improvement:
```scala
private static final ClassLoaderCache ANALYZING_COMPILER_CLASSLOADER_CACHE =
   new ClassLoaderCache(new URLClassLoader(new URL[0]));
/// ...
if (scalaCompiler instanceof AnalyzingCompiler) {
            scalaCompiler = ((AnalyzingCompiler) scalaCompiler).withClassLoaderCache(ANALYZING_COMPILER_CLASSLOADER_CACHE);
}
```
This is something I found while trying to speed up the compilation even further. Seems a little risky to include, as I haven't found any documentation about this cache usage.

IMO, FixV2 should be good for now. Further optimizations can be be added later.

**Comment by ijuma** [comment]
Cc @retronym in case he has thoughts.

**Comment by retronym** [comment]
As you've found, `sbt.internal.inc.AnalyzingCompiler` contains a [`classLoaderCache`](https://github.com/sbt/zinc/blob/6b357ae59a86a1238d78c4f808662e808747259b/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/AnalyzingCompiler.scala#L44-L57): [`ClassloaderCache`](https://github.com/sbt/zinc/blob/6b357ae59a86a1238d78c4f808662e808747259b/internal/zinc-classpath/src/main/scala/sbt/internal/inc/classpath/ClassLoaderCache.scala) field. Successive instantiations of `AnalyzingCompiler` can share an instance of this cache.

Depending on which factory methods you are using, it might be simpler to get the cache from the first compiler you create and then call `withClassLoaderCache` to assign that to subsequent instances.

[Here's](https://github.com/sbt/sbt/blob/38a857d887bf5148d9efa207a0ab88c16e8a5e57/main/src/main/scala/sbt/Defaults.scala#L542-L585) where SBT uses a process-wide instance of that cache when instantiating Zinc. It seems to use some non-standard factories, I'm not sure why.

The `ClassLoaderCache` should be `closed()` when disposing it your compiler wrappers to close cached classloaders.  This releases locks on files on Windows and allows the JVM to unload classes.

**Comment by eatkins** [comment]
More context on how sbt manages its `ClassLoaderCache` is available here: https://github.com/sbt/sbt/commit/03bf539ce977951d299456186d9ce5ccb735b2a9. The basic idea is that we want there to be one global `ClassLoaderCache` to minimize the classloading overhead and reduce the metaspace footprint of sbt. What I think is different from gradle is that sbt also evaluates `run` and `test` in process and the global `ClassLoaderCache` is also used for the classloaders used by those tasks but, if possible, it recycles layers that were used for compilation. As @retronym suggested, the sbt implementation of `ClassLoaderCache` has a `ReferenceQueue` of `ClassLoader` and a reaper thread to close those loaders that are no longer referenced.

**Comment by SheliakLyr** [comment]
Thank you all for the hints. If I understand correctly, I can use the ClassLoaderCache for different AnalysisCompiler instances. This seems to be the reason for it's existence.

After reading your comments, I understood that my current cache is already buggy: uses weak references instead of soft references (that's trivial) and does not close the constructed classloaders. Default implementation of a ClassLoaderCache (actually AbstractClassLoaderCache) is also leaky.

My idea right now is to create a simple implementation of classloader cache based on guava. I have 3 variants:

1. Llimited size + strong references + close on eviction.  Very simple, but risky if we can't get the max size right. On the other hand, those caches do not need to be very big, as I do not expect a lot of misses. Gradle worker is session-bound so this caches will be dropped after a single run of the build tool.
2. Limited size + soft references + reference queue + cleanup on `get()` operation. In this case, guava cache will be of type `Cache[Key, ClassLoaderReference]` where ClassLoaderReference is a subtype of `SoftReference`. This solution tries to avoid an additional thread, which unfortunately might be wrong. Cleaning reference queue on `get()` operations will effectively keep entire classloader in memory (`ClassLoaderReference` must keep a strong reference to  `Closeable` classloader and soft reference for just a delegating object). In the case of a memory pressure, soft values will be released, but this will still keep classloader in memory until `get()` is called.
3. Limited size + soft references + reference queue + reaper thread. Requires additional threads :(

Currently, I have an experimental version of the 2nd approach on branch cachedScalaCompilerV5. Speedup is significant (~28s). Personally however, I think that (1) should do the job. We can use very small cache sizes (5). Typical project will probably use the same classloader for all modules.

Guava'a soft/weak values cannot be used, as the lack o proper mechanism for resource release.

For ClassLoaderCache, as a key, I use a pair (filename, timestamp). This is not a gradle-way, but consistent with the way sbt works and @big-guy noticed that the hashing infrastructure is not available in the worker.

**Comment by SheliakLyr** [comment]
After some thinking, I decided to use the 1st approach from my previous post. Meaning no soft values, caches are limited only by a number of entries. This is less cumbersome and I personally don't have a very good experience with soft references (or rather... the way they impact gc).

I have just force-pushed a new version. Please review & tell me what you think.

File: subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java [hunk 1]
```diff
@@ -43,15 +52,107 @@
...
+        }
+    }
+
+    /**
+     * This class implements AbstractClassLoaderCache in a way that allows safe
+     * resource release when entries are evicted.
+     *
+     * Cache is based on file timestamps, because this method is used in sbt implementation.
+     */
+    static class TimeCheckingClassLoaderCache implements AbstractClassLoaderCache {
```
Comment (ljacomet) [inline subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java:104]: The `AbstractClassLoaderCache` does not exist in Zinc 1.2.0 which is supported by Gradle.

File: subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java [hunk 2]
```diff
@@ -62,39 +163,59 @@ static ZincScalaCompiler getCompiler(CacheRepository cacheRepository, final Iter
...
-        File compilerBridgeSourceJar = findFile("compiler-bridge", scalaClasspath);
+        File compilerBridgeSourceJar = findFile("compiler-bridge", hashedScalaClasspath.getClasspath());
         File bridgeJar = getBridgeJar(zincCache, scalaInstance, compilerBridgeSourceJar, sbt.util.Logger.xlog2Log(new SbtLoggerAdapter()));
-        ScalaCompiler scalaCompiler = ZincCompilerUtil.scalaCompiler(scalaInstance, bridgeJar, ClasspathOptionsUtil.auto());
+        ScalaCompiler scalaCompiler = new AnalyzingCompiler(
+            scalaInstance,
+            ZincUtil.constantBridgeProvider(scalaInstance, bridgeJar),
+            ClasspathOptionsUtil.auto(),
+            k -> scala.runtime.BoxedUnit.UNIT,
+            Option.apply(new ClassLoaderCache(COMPILER_CLASSLOADER_CACHE))
```
Comment (ljacomet) [inline subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java:173]: Since we cannot use the `AbstractClassLoaderCache`, what are the options here?
I admit not being 100% at ease with what this exactly means.
From basic testing, it seems that replacing `new ClassLoaderCache(COMPILER_CLASSLOADER_CACHE)` with `new ClassLoaderCache(scalaInstance.loader())` works (in the sense that it allows Gradle tests to pass) though I did not check the performance impact.

An alternative is to forget about the internal Zinc class loader cache and go back to invoking `ZincCompilerUtil.scalaCompiler`.

In short, the changes done should not make `UpToDateScalaCompileIntegrationTest` fail. Note that in that test the `@Requires` on the first test method can be removed.

Comment (SheliakLyr) [reply subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java:173]: Test shows that `ClassLoaderCache(scalaInstance.loader())` does not bring any noticeable gains over `ZincCompilerUtil.scalaCompiler`. The idea is pretty interesting, but it seems that compiler loads different classes than loader (bridge jar?). I also think that it is a little hackish to assume which classes AnalyzingCompiler will read.

Before we cross out AbstractClassLoaderCache... how about using reflection to load AbstractClassLoaderCache/TimeCheckingClassLoaderCache only when it is possible and otherwise use default ClassLoaderCache (which might leak some resources) or recreate it each time (which is slower by ~33%). By default Zinc 1.3.5 is used, so most of the users will benefit fully from the fix.

Comment (ijuma) [reply subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java]: Yeah, I think it makes a lot of sense to optimize for Zinc 1.3.5 since that's what is shipped by default. Is there a good reason to use older versions of Zinc?

**Comment by ljacomet** [comment]
@bot-gradle test RFM

File: subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java (hunk 2 above)
Comment (ljacomet) [reply subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java]: > Is there a good reason to use older versions of Zinc?

Not sure there, not a Scala user myself. However we try hard to not break compatibility in minors. We can certainly discuss dropping Zinc 1.2.x support for Gradle 7.0 for example.
For now, I think having the code handle both cases is best.

Let's see how the PR behaves through the full test suite.

Thanks again for being reactive and following up.

**Comment by ijuma** [comment]
This is great news. We updated Apache Kafka to Gradle 6.3 (and later Gradle 6.4) and there have been multiple complaints regarding compilation speed. So, I'm encouraged by this development.

**Comment by ljacomet** [comment]
This is now merged, thanks again!

Please let others know so that we get testing with 6.5 RC1 which should be out early next week.

**Comment by SheliakLyr** [comment]
Thanks @ljacomet! I am really glad you merged this already. Eagerly awaiting RC1.

**Comment by ijuma** [comment]
Looks like the milestone associated should be updated from 6.6 RC1 to 6.5 RC1.

**Comment by ljacomet** [comment]
🤦  Fixed. 
Looks like in my eagerness to include in 6.5, I forgot to update both the issue and the PR milestones.
